File: swift/grains/Sources/Grains/Grains.swift
enum Grains {
  enum GrainsError: Error {
    case inputTooHigh(String)
    case inputTooLow(String)

    static func forInput(_ input: Int) -> GrainsError {
      let message = "Input[\(input)] invalid. "
                  + "Input should be between 1 and 64 (inclusive)"

      return input < 1 ? inputTooLow(message) : inputTooHigh(message)
    }
  }

  static let total = UInt64.max

  static func square(_ square: Int) throws -> UInt {
    guard square >= 1 && square <= 64 else {
      throw GrainsError.forInput(square)
    }

    return UInt(1) << (square - 1)
  }
}

File: c/grains/src/grains.c
#include "grains.h"

__attribute__((pure)) u64 total(void) { return ~0ull; }
__attribute__((pure)) u64 square(const u8 i) {
  return (64 < i || i < 1) ? 0ull : 1ull << (i - 1);
}
